walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,181 @@
+      return Arrays.asList(TimeRange.fromStartEnd(START_OF_DAY, END_OF_DAY, false));
+    }
+
+    // First track times unavailable for the group
+    List<TimeRange> unavailableTimes = new ArrayList<>();
+    List<TimeRange> optionalUnavailableTimes = new ArrayList<>();
+    for (Event e : events) {
+      // If the attendee lists are not disjoint (meaning there is at least one attendee in common), 
+      // save this event time as unavailable
+      if (!Collections.disjoint(e.getAttendees(), mandatoryAttendees)) {
+        unavailableTimes.add(e.getWhen());
+      } 
+      if (!Collections.disjoint(e.getAttendees(), optionalAttendees)) {
+        optionalUnavailableTimes.add(e.getWhen());
+      } 
+    }
+
+    // Coalesce any overlapping times in the lists
+    coalesceList(unavailableTimes);
+    coalesceList(optionalUnavailableTimes);
+
+    List<TimeRange> availableTimes = new ArrayList<>();
+    List<TimeRange> optionalAvailableTimes = new ArrayList<>();
+    long duration = request.getDuration();
+
+    if (!unavailableTimes.isEmpty()) {
+      // Check if any times are open for mandatory attendees
+      // Methods add open times to availableTimes list
+      checkMorning(unavailableTimes, availableTimes, duration);
+      checkBetweenMeetings(unavailableTimes, availableTimes, duration);
+      checkEvening(unavailableTimes, availableTimes, duration);
+    }
+    else if (!mandatoryAttendees.isEmpty() && duration <= END_OF_DAY) {
+      // Entire day is open
+      availableTimes.add(TimeRange.fromStartEnd(START_OF_DAY, END_OF_DAY, false));
+    }
+
+    if (!optionalUnavailableTimes.isEmpty()) {
+      // Check if any times are open for optional attendees
+      // Methods add open times to optionalAvailableTimes list
+      checkMorning(optionalUnavailableTimes, optionalAvailableTimes, duration);
+      checkBetweenMeetings(optionalUnavailableTimes, optionalAvailableTimes, duration);
+      checkEvening(optionalUnavailableTimes, optionalAvailableTimes, duration);
+    }
+    else if (!optionalAttendees.isEmpty() && duration <= END_OF_DAY) {
+      // Entire day is open
+      optionalAvailableTimes.add(TimeRange.fromStartEnd(START_OF_DAY, END_OF_DAY, false));
+    }
+
+    // Attempt to combine to two lists to see if there are times when mandatory
+    // and optional attendees are both available
+    List<TimeRange> totalAvailableTimes = combineAvailableLists(availableTimes, optionalAvailableTimes, duration);
+
+    if (availableTimes.isEmpty()) {
+      return optionalAvailableTimes;
+    }
+    else if (totalAvailableTimes.isEmpty()) {
+      return availableTimes;
+    }
+    else {
+      return totalAvailableTimes;
+    }    
+  }
+
+ /**
+  * coalesceList: runs through a list of times and coalesces any
+  * overlapping times so that every entry in the list is a 
+  * distinct and separate time.
+  */
+  public static void coalesceList(List<TimeRange> times) {
```

Comment:
I would make this private; it's not really relevant on its own to the functionality of the class its in (aside from being an internal helper), and it's also pretty specialized, since it changes the list in place.

If this were to be a public interface, I'd move it out of this class into a separate utility class, and then I'd have it assert that the list is correctly sorted (and fail if it's not). I'd also likely have it create a copy, rather than change the list in place. THAT SAID, don't do this - I'm just saying what it'd take for it to warrant being public.